In [11]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from gensim.models import Word2Vec
import seaborn as sns
import nltk
import warnings 
warnings.filterwarnings("ignore", category=DeprecationWarning)
import os

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.

Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
Исключим стоп-слова с помощью stop_words='english'.
Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().

In [2]:
processed_df = pd.read_pickle('result.pkl')
processed_df.head()

id  label                                              tweet  \
0   1    0.0   @user when a father is dysfunctional and is s...   
1   2    0.0  @user @user thanks for #lyft credit i can't us...   
2   3    0.0                                bihday your majesty   
3   4    0.0  #model   i love u take with u all the time in ...   
4   5    0.0             factsguide: society now    #motivation   

                                         clean_tweet  \
0  when father is dysfunctional and is so selfish...   
1  thanks for lyft credit cannot use cause they d...   
2                                bihday your majesty   
3    model love you take with you all the time in ur   
4                  factsguide society now motivation   

                                         tweet_token  \
0  [when, father, is, dysfunctional, and, is, so,...   
1  [thanks, for, lyft, credit, can, not, use, cau...   
2                            [bihday, your, majesty]   
3  [model, love, you, take, with, you, all, the, ...   
4             [factsguide, society, now, motivation]   

                                tweet_token_filtered  \
0  [father, dysfunctional, selfish, drags, kids, ...   
1  [thanks, lyft, credit, use, cause, offer, whee...   
2                                  [bihday, majesty]   
3                      [model, love, take, time, ur]   
4                  [factsguide, society, motivation]   

                                       tweet_stemmed  \
0  [father, dysfunct, selfish, drag, kid, dysfunc...   
1  [thank, lyft, credit, use, caus, offer, wheelc...   
2                                  [bihday, majesti]   
3                      [model, love, take, time, ur]   
4                        [factsguid, societi, motiv]   

                                    tweet_lemmatized  
0  [father, dysfunctional, selfish, drag, kid, dy...  
1  [thank, lyft, credit, use, cause, offer, wheel...  
2                                  [bihday, majesty]  
3                      [model, love, take, time, ur]  
4                  [factsguide, society, motivation]

In [5]:
vectorizer = CountVectorizer(max_df=0.9, stop_words='english', max_features=1000)
stemmed = processed_df['tweet_stemmed'].apply(lambda x: ' '.join(x))
BOW_stemmed = vectorizer.fit_transform(stemmed)
feature_names = vectorizer.get_feature_names()
pd.DataFrame(BOW_stemmed.toarray(), columns = feature_names)

abl  absolut  accept  account  act  action  actor  actual  ad  adapt  \
0        0        0       0        0    0       0      0       0   0      0   
1        0        0       0        0    0       0      0       0   0      0   
2        0        0       0        0    0       0      0       0   0      0   
3        0        0       0        0    0       0      0       0   0      0   
4        0        0       0        0    0       0      0       0   0      0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ..    ...   
49154    0        0       0        0    0       0      0       0   0      0   
49155    0        0       0        0    0       0      0       0   0      0   
49156    0        0       0        0    0       0      0       0   0      0   
49157    0        0       0        0    0       0      0       0   0      0   
49158    0        0       0        0    0       0      0       0   0      0   

       ...  yeah  year  yesterday  yo  yoga  york  young  youtub  yr  yummi  
0      ...     0     0          0   0     0     0      0       0   0      0  
1      ...     0     0          0   0     0     0      0       0   0      0  
2      ...     0     0          0   0     0     0      0       0   0      0  
3      ...     0     0          0   0     0     0      0       0   0      0  
4      ...     0     0          0   0     0     0      0       0   0      0  
...    ...   ...   ...        ...  ..   ...   ...    ...     ...  ..    ...  
49154  ...     0     0          0   0     0     0      0       0   0      0  
49155  ...     0     0          0   0     0     0      0       0   0      0  
49156  ...     0     0          0   0     0     0      0       0   0      0  
49157  ...     0     0          0   0     0     0      0       0   0      0  
49158  ...     0     0          0   0     0     0      0       0   0      0  

[49159 rows x 1000 columns]

In [6]:
vectorizer = CountVectorizer(max_df=0.9, stop_words='english', max_features=1000)
lemmatized = processed_df['tweet_lemmatized'].apply(lambda x: ' '.join(x))
BOW_lemmatized = vectorizer.fit_transform(lemmatized)
feature_names = vectorizer.get_feature_names()
pd.DataFrame(BOW_lemmatized.toarray(), columns = feature_names)

able  absolutely  accept  account  act  action  actor  actually  adapt  \
0         0           0       0        0    0       0      0         0      0   
1         0           0       0        0    0       0      0         0      0   
2         0           0       0        0    0       0      0         0      0   
3         0           0       0        0    0       0      0         0      0   
4         0           0       0        0    0       0      0         0      0   
...     ...         ...     ...      ...  ...     ...    ...       ...    ...   
49154     0           0       0        0    0       0      0         0      0   
49155     0           0       0        0    0       0      0         0      0   
49156     0           0       0        0    0       0      0         0      0   
49157     0           0       0        0    0       0      0         0      0   
49158     0           0       0        0    0       0      0         0      0   

       add  ...  yesterday  yo  yoga  york  young  youth  youtube  yr  yrs  \
0        0  ...          0   0     0     0      0      0        0   0    0   
1        0  ...          0   0     0     0      0      0        0   0    0   
2        0  ...          0   0     0     0      0      0        0   0    0   
3        0  ...          0   0     0     0      0      0        0   0    0   
4        0  ...          0   0     0     0      0      0        0   0    0   
...    ...  ...        ...  ..   ...   ...    ...    ...      ...  ..  ...   
49154    0  ...          0   0     0     0      0      0        0   0    0   
49155    0  ...          0   0     0     0      0      0        0   0    0   
49156    0  ...          0   0     0     0      0      0        0   0    0   
49157    0  ...          0   0     0     0      0      0        0   0    0   
49158    0  ...          0   0     0     0      0      0        0   0    0   

       yummy  
0          0  
1          0  
2          0  
3          0  
4          0  
...      ...  
49154      0  
49155      0  
49156      0  
49157      0  
49158      0  

[49159 rows x 1000 columns]

Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.

Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
Исключим стоп-слова с помощью stop_words='english'.
Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().

In [7]:
vectorizer = TfidfVectorizer(max_df=0.9, stop_words='english', max_features=1000)
BOW_stemmed = vectorizer.fit_transform(stemmed)
feature_names = vectorizer.get_feature_names()
pd.DataFrame(BOW_stemmed.toarray(), columns = feature_names)

abl  absolut  accept  account  act  action  actor  actual   ad  adapt  \
0      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
1      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
2      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
3      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
4      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ...    ...   
49154  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49155  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49156  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49157  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49158  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   

       ...  yeah  year  yesterday   yo  yoga  york  young  youtub   yr  yummi  
0      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
1      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
2      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
3      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
4      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
...    ...   ...   ...        ...  ...   ...   ...    ...     ...  ...    ...  
49154  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49155  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49156  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49157  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49158  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  

[49159 rows x 1000 columns]

In [8]:
vectorizer = TfidfVectorizer(max_df=0.9, stop_words='english', max_features=1000)
BOW_lemmatized = vectorizer.fit_transform(lemmatized)
feature_names = vectorizer.get_feature_names()
pd.DataFrame(BOW_lemmatized.toarray(), columns = feature_names)

able  absolutely  accept  account  act  action  actor  actually  adapt  \
0       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
1       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
2       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
3       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
4       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
...     ...         ...     ...      ...  ...     ...    ...       ...    ...   
49154   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49155   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49156   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49157   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49158   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   

       add  ...  yesterday   yo  yoga  york  young  youth  youtube   yr  yrs  \
0      0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   
1      0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   
2      0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   
3      0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   
4      0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   
...    ...  ...        ...  ...   ...   ...    ...    ...      ...  ...  ...   
49154  0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   
49155  0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   
49156  0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   
49157  0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   
49158  0.0  ...        0.0  0.0   0.0   0.0    0.0    0.0      0.0  0.0  0.0   

       yummy  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0  
...      ...  
49154    0.0  
49155    0.0  
49156    0.0  
49157    0.0  
49158    0.0  

[49159 rows x 1000 columns]

Натренируем gensim.models.Word2Vec модель на наших данных.

●Тренировать будем на токенизированных твитах combine_df['tweet_token']
●Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
●Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.

In [12]:
X_train = processed_df['tweet_token'].tolist()

model = Word2Vec(
    X_train, 
    size=200, 
    window=5, 
    min_count=1, 
    sg = 1, 
    hs = 0, 
    negative = 10, 
    workers= 32, 
    seed = 34)

total_examples = len(X_train)
model.train([["hello", "world"]], total_examples=total_examples, epochs=20)

(40, 40)

Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

In [13]:
model.most_similar(positive = "dinner")

[('brunch', 0.9059646129608154),
 ('lunch', 0.9010446667671204),
 ('bestie', 0.8962730169296265),
 ('desse', 0.8921156525611877),
 ('tea', 0.8874828815460205),
 ('picnic', 0.8842434883117676),
 ('sushi', 0.8840305209159851),
 ('familytime', 0.8804506063461304),
 ('prom', 0.8796421885490417),
 ('bbq', 0.8775848150253296)]

In [14]:
model.most_similar(positive = "trump")

[('republican', 0.8027400970458984),
 ('blm', 0.7980953454971313),
 ('fuhered', 0.7966358661651611),
 ('remarks', 0.7961505651473999),
 ('clinton', 0.7920836806297302),
 ('paladino', 0.790328323841095),
 ('gop', 0.7889735698699951),
 ('carl', 0.7867719531059265),
 ('ally', 0.7860608100891113),
 ('campaign', 0.7832604646682739)]

Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
Давайте проверим векторное представление любого слова из нашего корпуса, например "food".

In [15]:
model.wv['food']

array([-5.46036303e-01,  5.48405871e-02,  1.35390803e-01, -3.03246807e-02,
       -1.09908484e-01,  1.91454232e-01,  5.29577553e-01, -1.58617318e-01,
       -5.32320142e-01, -6.02840543e-01, -4.99629974e-03, -5.79697073e-01,
        1.26181960e-01, -3.81314129e-01,  1.12527363e-01,  2.43439317e-01,
       -1.46169990e-01, -8.16847533e-02, -1.37027100e-01,  6.96618035e-02,
       -5.28773144e-02,  4.62823883e-02, -3.11656654e-01, -3.64486992e-01,
       -3.58587205e-01, -5.15616655e-01, -7.52681494e-02,  9.89614800e-02,
        5.10551214e-01, -2.07707450e-01, -5.21221310e-02, -1.85239330e-01,
       -2.94412643e-01, -6.86091065e-01,  5.87700009e-01,  7.96562374e-01,
       -2.51197577e-01, -2.95257092e-01,  2.79889852e-01,  1.15019411e-01,
        3.68225947e-02,  1.20731881e-02, -2.41941646e-01, -1.11278087e-01,
       -2.24438578e-01,  4.64705704e-03, -1.07437558e-03,  2.47670412e-01,
       -5.81486337e-02,  4.50611442e-01, -3.87858540e-01,  1.16879819e-03,
        1.08584641e-02, -

Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.

Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size))
и поделить финальный вектор на количество слов в твите.
На выходе должен получиться wordvec_df.shape = (49159, 200).

In [16]:
wordvec_df = np.zeros((total_examples, 200))
for i, tweet in enumerate(processed_df['tweet_token']):
    vec = np.zeros((1, 200))
    for word in tweet:
        vec += model.wv[word].reshape((1, 200))
    vec /= len(tweet)
    wordvec_df[i] = vec

D:\Prog\Anaconda3\envs\tensorflow\lib\site-packages\ipykernel_launcher.py:6: RuntimeWarning: invalid value encountered in true_divide
  


In [17]:
wordvec_df.shape

(49159, 200)